go-controller/pkg/ovn/address_set/address_set.go

```diff
@@ -74,45 +74,54 @@ func (asf *ovnAddressSetFactory) NewAddressSet(name string, ips []net.IP) (Addre
 	return res, nil
 }
 
-// ForEachAddressSet will pass the unhashed address set name, namespace name
-// and the first suffix in the name to the 'iteratorFn' for every address_set in
-// OVN. (Unhashed address set names are of the form namespaceName[.suffix1.suffix2. .suffixN])
-func (asf *ovnAddressSetFactory) ForEachAddressSet(iteratorFn AddressSetIterFunc) error {
+func forEachAddressSet(do func(string)) error {
 	output, stderr, err := util.RunOVNNbctl("--format=csv", "--data=bare", "--no-heading",
 		"--columns=external_ids", "find", "address_set")
 	if err != nil {
 		return fmt.Errorf("error reading address sets: "+
 			"stdout: %q, stderr: %q err: %v", output, stderr, err)
 	}
 
-	processedAddressSets := sets.String{}
 	for _, line := range strings.Split(output, "\n") {
 		for _, externalID := range strings.Split(line, ",") {
 			if !strings.HasPrefix(externalID, "name=") {
 				continue
 			}
-			// Remove the suffix from the address set name and normalize
-			addrSetName := truncateSuffixFromAddressSet(externalID[5:])
-			if processedAddressSets.Has(addrSetName) {
-				// We have already processed the address set. In case of dual stack we will have _v4 and _v6
-				// suffixes for address sets. Since we are normalizing these two address sets through this API
-				// we will process only one normalized address set name.
-				break
-			}
-			processedAddressSets.Insert(addrSetName)
-			names := strings.Split(addrSetName, ".")
-			addrSetNamespace := names[0]
-			nameSuffix := ""
-			if len(names) >= 2 {
-				nameSuffix = names[1]
-			}
-			iteratorFn(addrSetName, addrSetNamespace, nameSuffix)
+			name := externalID[5:]
+			do(name)
 			break
 		}
 	}
 	return nil
 }
 
+// ForEachAddressSet will pass the unhashed address set name, namespace name
+// and the first suffix in the name to the 'iteratorFn' for every address_set in
+// OVN. (Unhashed address set names are of the form namespaceName[.suffix1.suffix2. .suffixN])
+func (asf *ovnAddressSetFactory) ProcessEachAddressSet(iteratorFn AddressSetIterFunc) error {
+	processedAddressSets := sets.String{}
+	err := forEachAddressSet(func(name string) {
+		// Remove the suffix from the address set name and normalize
+		addrSetName := truncateSuffixFromAddressSet(name)
+		if processedAddressSets.Has(addrSetName) {
+			// We have already processed the address set. In case of dual stack we will have _v4 and _v6
+			// suffixes for address sets. Since we are normalizing these two address sets through this API
+			// we will process only one normalized address set name.
+			return
+		}
+		processedAddressSets.Insert(addrSetName)
+		names := strings.Split(addrSetName, ".")
+		addrSetNamespace := names[0]
+		nameSuffix := ""
+		if len(names) >= 2 {
```

Comment:
should this ever be greater than 2? isnt that an error?

Reply:
Yes, I don't think this can be grater than 2 at this time, but since this test is a guard to check if we can grab `names[1]` I think is good approach and future proof to do the least specific test necessary. It was like this before and I kinda agreed with it. Let me know if you would prefer to test for equality here.

Reply:
well if you had 3, then the suffix would be names[1] or names[2]? seems like we dont know because its an unexpected case, so we should handle it and throw an error if somehow we got an address set name that split into 3 parts. Or we could SplitN and always get 2 and assume whatever is after the first split is always the suffix.

Reply:
The method has this doc:
```
// ProcessEachAddressSet will pass the unhashed address set name, namespace name
// and the first suffix in the name to the 'iteratorFn' for every address_set in
// OVN. (Unhashed address set names are of the form namespaceName[.suffix1.suffix2. .suffixN])
```
So we can have N suffix and if we had 3 then we would still pass `names[1] ` as explained in the doc. If it were me I would pass a variadic argument with all the suffixes but that was how it was before and I had no reason to change it. We can do it though.

Reply:
ah i didnt relize we support multiple suffixes like that explicitly, nvm its fine